Hpcc 19154 Custom ESDL Request Transform implementation

rpastrana opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented:
@afishbeck this implements your xmllib xpath processor proposal, please provide informal high levelreview but note that libxml2 xmlhashTable was not successfully used to set $xmlvars  as designed. Also, ignore commented out code, it's there for a reason and will be removed before finalizing.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 1c98c18bf7b546afcc469b138d9f01e6bf096d99
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  173 sec (00:02:53) |   60 sec (00:01:00) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1192 sec (00:19:52) |   20 sec (00:00:20) | 1501 sec (00:25:01) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8064001926b26d2d218774bd4183872c746e70d4
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  173 sec (00:02:53) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1386 sec (00:23:06) |   19 sec (00:00:19) | 1692 sec (00:28:12) |

rpastrana commented:
@afishbeck please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 055b1593596318d34e28cad27c83f292141237d2
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  172 sec (00:02:52) |   60 sec (00:01:00) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1397 sec (00:23:17) |   19 sec (00:00:19) | 1715 sec (00:28:35) |

rpastrana commented:
@timothyklemm please review

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -509,6 +515,10 @@ void EsdlServiceImpl::handleServiceRequest(IEspContext &context,
         else
         {
+            context.addTraceSummaryTimeStamp(LogNormal, "srt-reqtrans");
+            m_customRequestTransform->processTransform(&context, req, m_oEspBindingCfg.get());
```
The req at this point hasn't been transformed in any way, applying the transform here allows clean xpath mapping into the req... but new fields resulting from the transform are subject for filtering in the esdl transform process line 527. Considering moving this transform post esdl transform. the request at that point will be wrapped in a row element...

afishbeck replied:
@rpastrana as we discussed, and as mentioned above this should happen after the built in ESDL transform.  This custom transform can be considered part of the service and should operate on already versioned data.  Also this transform should be able to add extra information for the next hop (usually Roxie) that isn't necessarily in the ESDL.  In the future we can consider adding some form of validation, etc.

afishbeck commented on esp/services/esdl_svc_engine/CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ set (   SRCS
         esdl_svc_engine.cpp
         esdl_store.cpp
         esdl_monitor.cpp
+        esdl_transform.cpp
```
This is a form of transform, but because ESDL is used to do so many different types of transforms I find this file name confusing.. I would open this file looking for even the basic built in ESDL transforms.  Something like esdl_svc_custom might be better.  We can put all features related to "scripting" service behavior in there, including these custom request transforms.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -164,6 +164,8 @@ void EsdlServiceImpl::init(const IPropertyTree *cfg,
     xpath.appendf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]", process, service);
     IPropertyTree * srvcfg = cfg->queryPropTree(xpath.str());
+    xpath.setf("Software/EspProcess[@name=\"%s\"]/EspBinding[@service=\"%s\"]", process, service); //get this service's binding cfg
+    m_oEspBindingCfg.setown(cfg->queryPropTree(xpath.str()));
```
Is this ultimately pointing directly to a remote DALI Ptree?  If so I wonder if there are any locking issues, and whether we should make a copy of the tree rather think linking to it.

Also you are doing a queryPropTree which doesn't link and then assingin ownership via setown.  That could end up causing a fault.

rpastrana replied:
I don't think so, CEspConfig creates the tree from the configfile and passes it down

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -30,6 +31,7 @@
 #include "dautils.hpp"
 #include "esdl_store.hpp"
 #include "esdl_monitor.hpp"
+#include "esdl_transform.hpp"
```
This seems to be added twice?

timothyklemm commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -290,8 +292,12 @@ void EsdlServiceImpl::configureTargets(IPropertyTree *cfg, const char *service)
+
     if (target_cfg)
     {
+        IPropertyTree * customRequestTransform = cfg->queryPropTree("Definition/Methods/Method/hpcc:Transforms");
```
queryPropTree will, internally, find all matching elements and throw an exception if more than one is found. Is this the expected behavior?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a295f78557a00d9e9b734ac3edfcfc430dd2b15c
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  171 sec (00:02:51) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1278 sec (00:21:18) |   18 sec (00:00:18) | 1580 sec (00:26:20) |

rpastrana commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,331 @@
+    if(!m_xpathContext->addParameter("clientversion", ver.str()))
+        ERRLOG("Could not set custom transform variable: clientversion:'%s'", ver.str());
+
+    auto user = context->queryUser();
```
@timothyklemm  please comment on the use of the getPropertyIterator you exposed.

timothyklemm replied:
getPropertyIterator returns a new iterator instance, or NULL when no properties have been set. The returned value should either be 'Owned' or explicitly released. It must not be dereferenced without first checking for NULL.

afishbeck reviewed (changes requested):
@rpastrana  Looks like a lot of comments / changes, but in reality I think the functionality is very close.

afishbeck commented:
@rpastrana FYI, in my test I couldn't quite figure out where the current xml node was in the xpath for the request.

Absolute paths like 
              <xsl:when test="/RoxieEchoPersonInfoRequest/Row/Name/First='joe'" required="false">
or
              <xsl:when test="/*/Row/Name/First='joe'" required="false">

but I couldn't get the relative path
           <xsl:when test="Row/Name/First='joe'" required="false">
or similar to work.

Maybe something needs to be done to set the context position.  Not sure.

I really don't mind if that ends up being a "known issue", since it functions ok with absolute paths, but if so we can open a jira.

afishbeck commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    Owned<IPropertyTreeIterator> configParams = bindingCfg->getElements("Transform/Param");
```
@rpastrana also might want a different name than "Transform" here.  So that it makes it really obvious when these are used.  Maybe even though its long "CustomRequestTransform" would be better in that its totally obvious.  But very open to better ideas.

afishbeck commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,331 @@
+    bool result = false;
+    try
+    {
+        bool result = evaluate(context, request,bindingCfg);
```
I could use a little explanation of this arrangement where you evaluate and then pass the result to various types of children.  Might be ok, but a little confusing to me.  Let's discuss offline, then I'll comment more here.

afishbeck replied:
@rpastrana

rpastrana commented on esp/services/esdl_svc_engine/esdl_svc_custom.hpp:
```diff
@@ -0,0 +1,146 @@
+    void addTransformRule(CEsdlCustomTransformRule * fieldmapping, bool otherwise);
+    void addNestedTransformClause(CEsdlCustomTransformClause * nestedConditional, bool otherwise);
+    bool evaluate(IEspContext * context, IPropertyTree *request, IPropertyTree * bindingCfg);
+    void process(IEspContext * context, IPropertyTree *request, IPropertyTree * bindingCfg);private:
```
good catch

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -290,14 +289,30 @@ void EsdlServiceImpl::configureTargets(IPropertyTree *cfg, const char *service)
+        }
+        catch (IPTreeException *e)
+        {
+            ERRLOG("Encountered error while fetching \"hpcc:Transforms\" from service \"%s\" ESDL configuration. Ensure a single set of transforms is provided ", service);
```
Expecting this to be ambiguous xpaths, but will add exception message just in case anything else goes through

rpastrana commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,331 @@
+    StringBuffer xmlreq;
+    toXML(request, xmlreq);
+
+    m_xpathContext->setXmlDoc(xmlreq.str());
```
absolutely, I think this got lost in the shuffle

rpastrana commented on esp/services/esdl_svc_engine/esdl_svc_custom.hpp:
```diff
@@ -0,0 +1,146 @@
+private:
+    StringBuffer m_targetField;
+    StringBuffer m_xpathToValue;
+    const char * m_ruleName;
```
Keeping in order to be able to reference rules in future

rpastrana commented:
@afishbeck please review latest commit.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 00092d60e662f5e9854b1195192a034d3595b4f5
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  172 sec (00:02:52) |   62 sec (00:01:02) |    6 sec (00:00:06) |   27 sec (00:00:27) | 1277 sec (00:21:17) |   19 sec (00:00:19) | 1585 sec (00:26:25) |

rpastrana commented:
<Method name="FraudGovReport">
  <xsdl:CustomRequestTransforms>
   <xsdl:CustomRequestTransform name="myfirsttransform">
    <xsdl:choose>
     <xsdl:when required="true" test="$clientversion=1.1">
      <xsdl:SetValue name="bogustransform" target="Row/Options/IsIdentityTestRequest"  value="//Row/Options/AppendBest" optional="false"/>
      <xsdl:SetValue name="bogusoptionaltransform2" target="Row/Options/IsOnline" value="10" optional="true"/>
     </xsdl:when>
     <xsdl:otherwise>
      <xsdl:SetValue name="firstotherwise" target="Options/IsIdentityTestRequest" value="//Options/AppendBest"/>
     </xsdl:otherwise>
    </xsdl:choose>
   </xsdl:CustomRequestTransform>
  </xsdl:CustomRequestTransforms>
 </Method>

afishbeck reviewed (commented):
@rpastrana looks good.  I like the changes.  

As we discussed an additional example in /opt/HPCCSystems/examples/EsdlExample would help a lot.  A second very simple version of calling the roxie query and changing the input parameters would be good.

As we discussed the organization could be improved a bit, but we can save that for a later update.  Have you tested for memory leaks just in case?  I'll also ask Yanrui to review.

rpastrana commented:
@afishbeck please open a new issue for the example for 7.0.2

mayx commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,346 @@
+    bool result = false;
+    try
+    {
+        bool result = evaluate(xpathContext);
```
What's the purpose of the result variable? It's not being used. Should you change the return value of the method to bool?

mayx commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,346 @@
+{
+    m_xpathContext.set(getXpathContext());
+    Owned<IPropertyTreeIterator> transformsIter = cfg->getElements("xsdl:CustomRequestTransform");
+    {
```
What's the purpose of the extra layer of scope?

mayx commented on esp/services/esdl_svc_engine/esdl_svc_custom.cpp:
```diff
@@ -0,0 +1,346 @@
+            }
+        }
+
+        Owned<IPropertyTree> thereq = createPTreeFromXMLString(request.str());
```
Doing string->ptree->string conversions twice hurts the performance quite significantly. Any chance the transform can be done before esdltransformer? Or maybe add a form of esdl transform that will keep the transformed result in a ptree format (under the condition of no performance penalty)?

rpastrana replied:
Very valid comment regarding the representation and transformation of the request from str to ptree back to str. Delayed the transform as long as possible. At the point we receive the req for custom transform, it's already been processed by esdl engine and returned as STR. We transform (late) the req into a tree in order to perform custom transform.

mayx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -0,0 +1,305 @@
+    ~CLibXpathContext()
+    {
+        xmlFreeDoc(m_xmlDoc);
+        xmlXPathFreeContext(m_xpathContext);
```
Any need to check for nullptr?

rpastrana replied:
function checks for nullptr

mayx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -0,0 +1,305 @@
+
+    static void tableScanCallback(void *payload, void *data, xmlChar *name)
+    {
+        DBGLOG("k/v == [%s,%s]\n", (char *) name, (char *) payload);
```
Would this generate a lot of tracing?

rpastrana replied:
not used, method available for debug

mayx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -0,0 +1,305 @@
+        if (xmldoc && * xmldoc)
+        {
+            WriteLockBlock wblock(m_rwlock);
+            m_xmlDoc = xmlParseDoc((const unsigned char *)xmldoc);
```
Should check if m_xmlDoc already exists, if yes, release first.

mayx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -0,0 +1,305 @@
+            }
+
+            // Create xpath evaluation context
+            m_xpathContext = xmlXPathNewContext(m_xmlDoc);
```
Similar as above.

mayx commented on system/xmllib/libxml_xpathprocessor.cpp:
```diff
@@ -0,0 +1,305 @@
+class CLibXpathContext : public CInterface, public IXpathContext
+{
+private:
+    xmlDocPtr m_xmlDoc = nullptr;
```
The xpath context is shared by all threads of the same binding, so the XmlDocPtr can not be a member variable. One thread sets it, another thread may set it to a different doc, while the first thread is still using it. You should pass this in as a parameter to a method, or create another class to contain it that will have an object instantiated for every request/thread.

mayx reviewed (commented):
@rpastrana @afishbeck some comments.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8fb065e50cb95a1a73e931486bdad46cecf145b8
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  173 sec (00:02:53) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1302 sec (00:21:42) |   20 sec (00:00:20) | 1605 sec (00:26:45) |

rpastrana reviewed (commented):
@mayx please take a look at latest commit.
You had a valid comment regarding the representation and transformation of the request from str to ptree back to str. Delayed the transform as long as possible.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6eb227def0cdefc4e625254b42bc1c11af25d807
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  170 sec (00:02:50) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1269 sec (00:21:09) |   17 sec (00:00:17) | 1572 sec (00:26:12) |

rpastrana commented:
@wangkx please review latest commit (both Tony and Yanrui are out).

wangkx commented:
@rpastrana the changes look good. Please add your comments as we discussed.

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -83,7 +85,7 @@ class EsdlServiceImpl : public CInterface, implements IEspService
     StringBuffer                m_espServiceName;
     StringBuffer                m_espProcName;
-    Owned<IPropertyTree>        m_pServiceConfig;
+    Owned<IPropertyTree>        m_oEspBindingCfg;
```
done, service config used locally

rpastrana committed:
HPCC-19154 DESDL Custom request transform implementation

- Create xmllib based xpathprocessor
- Utilize xpathprocessor to evaluate xsl:when test attribute
- Capture binding config for consumption by template transform
- Support nested conditional transforms

- Adds explicit error messages when non-optional rules can't be evaluated
- Avoids processing invalid transforms

- Restructures getXpathContext for thread safety
- Renames IXpathContext::addParameter to addVariable for consistency
- Replaces use of xmlXPathWrapCString with xmlXPathNewCString for mem management
- All reqs with custom transform creates temporary xpath obj
- Adds flag to reject request if custom transform template compile failed
- Re-introduces use of services' config in init

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@richardkchapman squashed. Please merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 661c196808cae00531b3a7a4c55bb7219e75abe2
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  173 sec (00:02:53) |   62 sec (00:01:02) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1334 sec (00:22:14) |   21 sec (00:00:21) | 1646 sec (00:27:26) |

richardkchapman commented:
@rpastrana  Should the commit message have WIP in the title?

rpastrana commented:
@richardkchapman commit message amended
